File: 2020-11-02/llista-primers.cc
// calcula la llista de tots els nombres primers fins a un valor donat

#include <iostream>
#include <vector>
using namespace std;

// retorna un vector p amb n posicions
// tal que p[i] és cert ssi i és primer
vector<bool> eratostenes(int n)
{
    vector<bool> p(n + 1, true);
    p[0] = false;
    p[1] = false;
    for (int i = 2; i * i <= n; ++i) {
        if (p[i]) {
            for (int j = 2 * i; j <= n; j += i) {
                p[j] = false;
            }
        }
    }
    return p;
}

// retorna un vector ordenat amb tots els nombres primers <= n.
vector<int> trobar_primers(int n)
{
    vector<int> primers;
    vector<bool> p = eratostenes(n);
    for (int i = 0; i < n; ++i) {
        if (p[i]) {
            primers.push_back(i);
        }
    }
    return primers;
}

int main()
{
    const int n = 20;
    vector<int> primers = trobar_primers(n);
    for (int primer : primers) {
        cout << primer << endl;
    }
}

File: 2020-11-02/llista-primers.py
# calcula la llista de tots els nombres primers fins a un valor donat

def eratostenes(n):
    p = [True for i in range(n + 1)]
    p[0] = False
    p[1] = False
    i = 2
    while i * i <= n:
        if p[i]:
            for j in range(2 * i, n + 1, i):
                p[j] = False
        i += 1
    return p


def llista_primers(n):
    p = eratostenes(n)
    primers = []
    for i in range(len(p)):
        if p[i]:
            primers.append(i)
    return primers


print(llista_primers(20))
